engine/src/main/java/org/terasology/entitySystem/entity/internal/PojoEntityCache.java

```diff
@@ -0,0 +1,350 @@
+            BeforeEntityCreated event = new BeforeEntityCreated(prefab, components);
+            BaseEntityRef tempRef = entityManager.getEntityRefStrategy().createRefFor(entityId, entityManager);
+            entityManager.getEventSystem().send(tempRef, event);
+            tempRef.invalidate();
+            finalComponents = event.getResultComponents();
+        } else {
+            finalComponents = components;
+        }
+
+        for (Component c : finalComponents) {
+            componentStore.put(entityId, c);
+        }
+        return createEntityRef(entityId);
+    }
+
+    @Override
+    public EntityRef createEntityRefWithId(long id) {
+        if (entityManager.isExistingEntity(id)) {
+            return createEntityRef(id);
+        }
+        return EntityRef.NULL;
+    }
+
+    /**
+     * Creates the entity without sending any events. The entity life cycle subscriber will however be informed.
+     */
+    @Override
+    public EntityRef createEntityWithoutLifecycleEvents(Iterable<Component> components) {
+        EntityRef entity = createEntity(components);
+        for (Component component: components) {
+            entityManager.notifyComponentAdded(entity, component.getClass());
+        }
+        return entity;
+    }
+
+    /**
+     * Creates the entity without sending any events. The entity life cycle subscriber will however be informed.
+     */
+    @Override
+    public EntityRef createEntityWithoutLifecycleEvents(String prefabName) {
+        return createEntityWithoutLifecycleEvents(entityManager.getPrefabManager().getPrefab(prefabName));
+    }
+
+    /**
+     * Creates the entity without sending any events. The entity life cycle subscriber will however be informed.
+     */
+    @Override
+    public EntityRef createEntityWithoutLifecycleEvents(Prefab prefab) {
+        if (prefab != null) {
+            List<Component> components = Lists.newArrayList();
+            for (Component component : prefab.iterateComponents()) {
+                components.add(entityManager.getComponentLibrary().copy(component));
+            }
+            components.add(new EntityInfoComponent(prefab, prefab.isPersisted(), prefab.isAlwaysRelevant()));
+
+            return createEntityWithoutLifecycleEvents(components);
+        } else {
+            return createEntityWithoutLifecycleEvents(Collections.<Component>emptyList());
+        }
+    }
+
+    /**
+     * Destroys the entity without sending any events. The entity life cycle subscriber will however be informed.
+     */
+    @Override
+    public void destroyEntityWithoutEvents(EntityRef entity) {
+        if (entity.isActive()) {
+            entityManager.notifyComponentRemovalAndEntityDestruction(entity.getId(), entity);
+            destroy(entity);
+        }
+    }
+
+    @Override
+    public EntityRef createEntityWithId(long id, Iterable<Component> components) {
+        if (!entityManager.registerId(id)) {
+            return EntityRef.NULL;
```

Comment:
An error gets logged by `registerId`. I'm not sure if it happens often or if it's a problem; this is only currently called for storage/serialization tasks, which already know the ID they're trying to create for from the entity they already had (so presumably they know that it's not invalid).

Most uses would only use the normal create methods, which will guarantee a valid ID. Perhaps a note should be added about this to the documentation?